protocol/directory_test.go

```diff
@@ -138,3 +139,231 @@ func TestRegisterExistedUserWithoutTB(t *testing.T) {
+	// expect a proof of absence
+	// and error ErrorNameNotFound
+	res, err = d.KeyLookup(&KeyLookupRequest{Username: "bob"})
+	if err != ErrorNameNotFound ||
+		res.Error != ErrorNameNotFound {
+		t.Fatal("Expect error code", ErrorNameNotFound, "got", err)
+	}
+	if ap := res.AP; ap == nil || bytes.Equal(ap.LookupIndex, ap.Leaf.Index) {
+		t.Fatal("Expect a proof of absence")
+	}
+	if res.TB != nil {
+		t.Fatal("Expect returned TB is nil")
+	}
+}
+
+func TestKeyLookupWithoutTB(t *testing.T) {
+	d := newDirectory(t, false)
+	_, err := d.Register(&RegistrationRequest{
+		Username: "alice",
+		Key:      []byte("key")})
+	if err != Success {
+		t.Fatal("Unable to register")
+	}
+	// lookup in the same epoch
+	// expect a proof of absence
+	// and error ErrorNameNotFound
+	res, err := d.KeyLookup(&KeyLookupRequest{Username: "alice"})
+	if err != ErrorNameNotFound ||
+		res.Error != ErrorNameNotFound {
+		t.Fatal("Expect error code", ErrorNameNotFound, "got", err)
+	}
+	if ap := res.AP; ap == nil || bytes.Equal(ap.LookupIndex, ap.Leaf.Index) {
+		t.Fatal("Expect a proof of absence")
+	}
+	if res.TB != nil {
+		t.Fatal("Expect returned TB is nil")
+	}
+
+	d.Update()
+	// lookup in epoch after registering epoch
+	// expect a proof of inclusion
+	res, err = d.KeyLookup(&KeyLookupRequest{Username: "alice"})
+	if ap := res.AP; ap == nil || !bytes.Equal(ap.LookupIndex, ap.Leaf.Index) {
+		t.Fatal("Expect a proof of inclusion")
+	}
+	if res.TB != nil {
+		t.Fatal("Expect returned TB is nil")
+	}
+}
+
+func TestKeyLookupNameNotFoundWithoutTB(t *testing.T) {
+	d := newDirectory(t, false)
+	_, err := d.Register(&RegistrationRequest{
+		Username: "alice",
+		Key:      []byte("key")})
+	if err != Success {
+		t.Fatal("Unable to register")
+	}
+	// lookup in the same epoch
+	// expect a proof of absence
+	res, err := d.KeyLookup(&KeyLookupRequest{Username: "alice"})
+	if err != ErrorNameNotFound ||
+		res.Error != ErrorNameNotFound {
+		t.Fatal("Expect error code", ErrorNameNotFound, "got", err)
+	}
+	if ap := res.AP; ap == nil || bytes.Equal(ap.LookupIndex, ap.Leaf.Index) {
+		t.Fatal("Expect a proof of absence")
+	}
+	if res.TB != nil {
+		t.Fatal("Expect returned TB is nil")
+	}
+
+	d.Update()
+	// lookup in different epochs
+	// expect a proof of absence
+	res, err = d.KeyLookup(&KeyLookupRequest{Username: "bob"})
+	if err != ErrorNameNotFound ||
+		res.Error != ErrorNameNotFound {
+		t.Fatal("Expect error code", ErrorNameNotFound, "got", err)
+	}
+	if ap := res.AP; ap == nil || bytes.Equal(ap.LookupIndex, ap.Leaf.Index) {
+		t.Fatal("Expect a proof of absence")
+	}
+	if res.TB != nil {
+		t.Fatal("Expect returned TB is nil")
+	}
+}
+
+func TestConsistencyCheck(t *testing.T) {
+	N := 10
+
+	d := newDirectory(t, false)
+	_, err := d.Register(&RegistrationRequest{
+		Username: "alice",
+		Key:      []byte("key")})
+	if err != Success {
+		t.Fatal("Unable to register")
+	}
+
+	d.Update()
+	savedSTR := d.LatestSTR().Signature
```

Comment:
It's fine to only save the STR signature here since it's a test, but in the client we should make sure to save a few more STR fields like the latestEpoch and the epochDeadline.